### PR TITLE
fix issue 16008 - FreeList should implement deallocateAll

### PR DESCRIPTION
**Pull request by BBasile**

### CODE REVIEW

File: std/experimental/allocator/building_blocks/free_list.d [hunk 1]
```diff
@@ -359,14 +359,27 @@ struct FreeList(ParentAllocator,
...
     /**
-    Defined only if $(D ParentAllocator) defines $(D deallocateAll). If so,
-    forwards to it and resets the freelist.
+    If the parent allocator implements `deallocateAll` then forwards it otherwise
+    each element is freed with the parent's `deallocate`.
     */
-    static if (hasMember!(ParentAllocator, "deallocateAll"))
     bool deallocateAll()
     {
+        bool result = false;
```
Comment (atilaneves) [inline std/experimental/allocator/building_blocks/free_list.d]: This could be moved to each static block:

```d
immutalbe result = parent.deallocateAll();
auto result = true;
//... set result to false
````

Comment (BBasile) [inline std/experimental/allocator/building_blocks/free_list.d]: I can do it but since the code is basically a paste from ShareFreeList sould I also patch it ?

File: std/experimental/allocator/building_blocks/free_list.d [hunk 2]
```diff
@@ -359,14 +359,27 @@ struct FreeList(ParentAllocator,
...
+        }
+        else static if (hasMember!(ParentAllocator, "deallocate"))
+        {
+            result = true;
+            for (auto n = root; n; n = n.next)
+            {
+                if (!parent.deallocate((cast(ubyte*)n)[0 .. max]))
+                    result = false;
+            }
+        }
```
Comment (ZombineDev) [inline std/experimental/allocator/building_blocks/free_list.d]: Shouldn't there be a `else static assert(0, "Not supported") `, because otherwise you would get a memory leak?

Comment (BBasile) [inline std/experimental/allocator/building_blocks/free_list.d]: Added the static assert. One question I didn't asked before is why the hell there is no template used as constraint for the different allocators ?  
A bit like _isInputRange_ eg

```D
template isMinimalAllocator(T)
{
    enum isMinimalAllocator = /*static checks*/...
}
```

Comment (DmitryOlshansky) [inline std/experimental/allocator/building_blocks/free_list.d]: Rather it should be in template constraint so that deallocateAll is not defined if the constraints are not met.

Comment (BBasile) [inline std/experimental/allocator/building_blocks/free_list.d]: There is no constraint at all in this package. Just look at the building blocks. I mean that it should be added everywhere and in another PR.

Comment (ZombineDev) [inline std/experimental/allocator/building_blocks/free_list.d]: Yeah, template constraints would be more true to the spirit of the design - providing functionality on best effort principle, yet with minimal requirements.

@BBasile if you have time, it would be best to turn the assert into a constraint, but if not we can leave this to a separate PR.

Comment (BBasile) [inline std/experimental/allocator/building_blocks/free_list.d]: Ok done, but what I was tried to say is that constraints could be added everywhere, not just to FreeList.

Comment (ZombineDev) [inline std/experimental/allocator/building_blocks/free_list.d]: Thanks. I agree they should, but let's take things one step at a time.

File: std/experimental/allocator/building_blocks/free_list.d [hunk 3]
```diff
@@ -3,6 +3,24 @@ module std.experimental.allocator.building_blocks.free_list;
...
 
+// indicates wether A is an allocator suitable to be used for a FreeList.
+private template isFreeListAlloc(A)
+{
+    enum isFreeListAlloc =
+        (is(A == class) || is(A == struct) || is(A == interface)) &&
+        __traits(hasMember, A, "allocate") &&
+        (__traits(hasMember, A, "deallocate") ||
+        __traits(hasMember, A, "deallocateAll"));
+}
```
Comment (ZombineDev) [inline std/experimental/allocator/building_blocks/free_list.d]: No. That's not what we had in mind. None of the allocators should have constraints like this. 
IAllocator has more than 10 members. If you took that road, you would need to make more than 2^10 constraints for every possible subset of methods which is clearly wrong.

Instead you should use the MDFINE pattern:
* The allocator *types* should not have any constraints, maybe except for `allocate`, which is (AFAIK) the only hard requirement. But since types can't be overloaded it should be a static assert and not a template constraint, because that way we can provide a better message. AFAIR most allocators assume that the parent allocator has `allocate` and that's fine becuase the compiler would complain otherwise.
* Every individual *method* should be *conditionally defined* (via static if (hasMethod...), and not restricted with a constraint), depending on whether the parent allocator can provide support for it.

---

MDFINE = Method Definition Failure Is Not an Error.

You can find more from Andrei's talks at CppCon and DConf 2015

File: std/experimental/allocator/building_blocks/free_list.d (hunk 2 above)
Comment (BBasile) [inline std/experimental/allocator/building_blocks/free_list.d]: Since a constraint is not useful I've finally made a static assert with a user friendly diagnostic.

**Comment by BBasile** [comment]
@braddr : you could have update the win32 auto-tester host. The time Zone for the Easter Island has changed and it looks like it's in the _Tuesday patch_ from the 10 May (new entry in the registry). Phobos tests have started to systematically fail because of this !!

> 
core.exception.AssertError@std\datetime.d(28135): TZName which is missing: Easter Island Standard Time

https://blogs.technet.microsoft.com/dst2007/2016/03/23/time-zone-update-for-chile-and-haiti/

**Comment by BBasile** [comment]
@braddr : please ignore any previous notification from mine. It's actually std.datetime that need to be patched not the host. Sorry.
